### ISSUE TITLE
Reconnecting a bot to a voice channel causes no audio to be heard from new dispatchers

### ISSUE THREAD

**Issue by ToppleKek** [discordjs/discord.js#2443]
**Please describe the problem you are having in as much detail as possible:**
While there is a dispatcher playing audio, after disconnecting the bot from the voice channel with `VoiceChannel.leave()` or `VoiceConnection.disconnect()` and reconnecting the bot to the voice channel with `VoiceChannel.join()` any new dispatchers will show on discord that they are sending audio, but nothing can be heard (Green ring, voice debug shows Audio Detected: Yes, and packets are being received) I'm pretty sure this isn't a client issue as I've ran other bots (other libs) on the same bot account that do not have this problem and a restart of discord does not solve this problem. Restarting the bot fixes this issue. Also, this happens per guild, if this problem occurs on guild A, the bot works fine on guild B and so on. I'm not sure if there's something I'm doing wrong but I've asked in both Discord.js Official and Discord API servers with no answers so I thought I'd write up a quick issue in the event that this is a library bug. Thanks in advance!

**Include a reproducible code sample here, if possible:**
```js
// For simplicity, lets say this bit is in a 'play' command.
const dispatcher = connection.play("./test.mp3", {
  volume: 1,
  passes: 3,
  bitrate: 96
});
// Again for simplicity, lets say this bit is in a 'summon' command.
const voiceConnection = msg.member.voiceChannel.join().catch(err => console.log(err));

// Yet again, lets say this bit is in a 'disconnect' command.
const connection = msg.guild.voiceConnection;
if (connection.dispatcher) connection.dispatcher.end(); // I have tried with and without this, both have the same issue.
connection.channel.leave();
// OR (Both have the same issue)
connection.disconnect();
```

**Further details:**

- discord.js version: v12 (master commit: e431ccd)
- node.js version: 8.10.0
- Operating system: Debian Unstable (sid)
- Priority this issue should have – please be realistic and elaborate if possible: Medium, extremely annoying, but not a huge issue that must be taken care of immediately.  

- [ ] I found this issue while running code on a __user account__
- [x] I have also tested the issue on latest master, commit hash: e431ccd

**Comment by hydrabolt** [comment]
It might be because the sequencing information is lost upon disconnect, I'll check this out tomorrow 👍

**Comment by ToppleKek** [comment]
Alright, thanks for the quick response :ok_hand:

**Comment by hydrabolt** [comment]
I'm unable to reproduce this, can you tell me how far apart your disconnect and reconnects roughly are (minutes, hours...) and what other libs you've tried that seem to be unaffected by this?

**Comment by hydrabolt** [comment]
That would be great! Thanks

**Comment by ToppleKek** [comment]
Alright, so in the first part of the video, I show that everything works just fine until I disconnect the bot (playing audio from a file btw) In the second part I show refreshing discord does not solve the problem. In the last bit I restarted the bot and played the same song, everything works. If I use the skip command (which is just ending the dispatcher) then disconnecting and reconnecting everything still works fine.

https://youtu.be/TCpq2r1MCm8

Could be something I'm doing wrong, but I don't know what that would be...

(Also forgot to mention that I've used a discord.py bot on this account with no issues)

**Comment by hydrabolt** [comment]
Thanks for bringing up this issue!

I was able to reproduce this issue eventually, and I _think_ I've now fixed it. Looks like Discord [added nonces](https://discordapp.com/developers/docs/topics/voice-connections#heartbeating) to the heartbeat packet and we didn't implement it; this must have slipped past me.

After I made sure we added a random nonce, it seems to have fixed the issue. If you could test again that'd be great!

**Comment by ToppleKek** [comment]
No problem! But, I've double checked to make sure that I've updated and the issue still persists. Yesterday after I uploaded the video I found a little work around that works, but probably isn't the best. I found that ending the dispatcher (if it exists) then using setTimeout to leave 100ms after that, everything works fine (Guessing the dispatcher takes a bit to stop speaking?) If it's fixed for you I suppose it could be an issue on my end... Let me know if there's anything else I should try/test, thanks!

**Comment by hydrabolt** [comment]
Unfortunately I can't reproduce this (I tried a lot!) I made a voice fix that _might_ be related, so if you could update and test again that'd be great.

Could you also send your code so I can verify that it it's valid?

Thanks!

**Comment by ToppleKek** [comment]
I really appreciate that you tried a lot to reproduce it! But sadly after updating it wasn't fixed on my end.
Here would be my disconnect command (Sloppy but whatever)
```js
  if (!msg.guild) return;
  if (!utils.checkPermission(msg.author, msg, "admin")) {
    utils.sendResponse(msg, "You do not have permission to use that command. You are missing permission `ADMINISTRATOR`", "err");
    return;
  }
  if (!msg.guild.voiceConnection) {
    utils.sendResponse(msg, "The bot is not in a voice channel", "err");
    return;
  }
  const connection = msg.guild.voiceConnection;
  queue[msg.guild.id].splice(1);
  if (connection.dispatcher) connection.dispatcher.end();
  connection.channel.leave();

  // This was my little work around
  //setTimeout(function () {
  // This is just some queue stuff
  //  queue[msg.guild.id].playing = false;
  //  queue[msg.guild.id].queueing = false;
  //  queue[msg.guild.id].firstSong = true;
  //  queue[msg.guild.id].loop = false;
  //  connection.channel.leave();
  //}, 100);
};
```

Here is my summon command:
```js
  if (!msg.guild) return;
  if (msg.member.voiceChannel) {
    if (!queue[msg.guild.id]) {
      // More queue stuff
      queue[msg.guild.id] = [];
      queue[msg.guild.id].playing = false;
      queue[msg.guild.id].queueing = false;
      queue[msg.guild.id].firstSong = true;
      queue[msg.guild.id].loop = false;
      queue[msg.guild.id].volume = 50;
      queue[msg.guild.id].searching = [];
    }
    const voiceConnection = msg.member.voiceChannel.join().catch(err => console.log(err));
  }
  else {
    utils.sendResponse(msg, "You must be in a voice channel!", "err");
  }
};
```
If nothing looks wrong, could I try your code on my machine? Thanks!

**Comment by hydrabolt** [comment]
Your code looks fine to me, I use test/voice.js to test (it's very badly written so I'll get back with a better sample script for you to use tomorrow)

Thanks!

**Comment by ToppleKek** [comment]
No need! I've found the problem although I'm not quite sure how. 

When the dispatcher finished, it would call a function to play the next song which would check to see if there was no voice connection first. If there wasn't a voice connection (`!msg.guild.voiceConnection`) it would reset the queue and `return;` I guess something there was breaking everything because after commenting it out everything works fine. Not quite sure what it is, if someone knows I'd love to hear.

Sorry to waste some of your time on trying to reproduce it again but I greatly appreciate the help and I guess we fixed the heartbeat issue ¯\_(ツ)_/¯

If there's anything else that needs to be done I can do so, thanks!
